asllib/ASLTypingReference.mld

```diff
@@ -722,124 +722,261 @@ of the following applies:
 
     {4 Rule}
+    All of the following applies:
+    - [s] is a for statement with index [id], direction [dir], two expressions
+      [e1] and [e2] and a statement [s'];
+    - [t1,e1'] is the result of annotating [e1] in [env];
+    - [t2,e2'] is the result of annotating [e2] in [env];
+    - [t1] has the structure of an integer type;
+    - [t2] has the structure of an integer type;
+    - [ty] is ;
+    - [s''] is the result of annotating [s'] in [env];
+    - [new_s] is a for statement with index [id], direction [dir], two expressions
+      [e1'] and [e2'] and statement [s''];
+    - [new_env] is [env].
 
     {4 Example}
 
-  {3 TypingRule.SThrowNone}
+  {3 TypingRule.SDeclSome}
 
     {4 Rule}
+    All of the following applies:
+    - [s] is a declaration with [ldk], [ldi] and an optional expression [e_opt];
+    - [e_opt] gives some expression [e];
+    - [t_e,e'] is the result of annotating [e] in [env];
+    - [env', ldi'] is the result of declaring the local constants of [ldi] in [env];
+    - [new_s] is a declaration with [ldk], [ldi'] and an expression [e'];
+    - [new_env] is [env].
 
     {4 Example}
 
-  {3 TypingRule.SThrowSome}
+  {3 TypingRule.SDeclNone}
 
     {4 Rule}
+    All of the following applies:
+    - [s] is a declaration statement with [ldk], [ldi] and an optional expression [e_opt];
+    - [e_opt] gives no expression;
+    - [env', s'] is the result of annotating uninitialised local declarations [ldi] in [env];
+    - [new_s] is [s'];
+    - [new_env] is [env].
 
     {4 Example}
-    <minimal example>
 
-  {3 TypingRule.STry}
+  {3 TypingRule.SThrowNone}
 
     {4 Rule}
+    All of the following applies:
+    - [s] is a throw statement with no expression;
+    - [new_s] is [s];
+    - [new_env] is [env].
 
     {4 Example}
 
-  {3 TypingRule.SDeclSome}
+  {3 TypingRule.SThrowSome}
 
     {4 Rule}
+    All of the following applies:
+    - [s] is a throw statement with expression [e];
+    - [t_e,e'] is the result of annotating [e] in [env];
+    - [t_e] has the structure of an exception type;
+    - [new_s] is a throw statement with expression [e'] and type [t_e];
+    - [new_env] is [env].
 
     {4 Example}
 
-  {3 TypingRule.SDeclNone}
+  {3 TypingRule.STry}
 
     {4 Rule}
+    All of the following applies:
+    - [s] is a try statement with statement [s'], catchers [catchers] and block [otherwise];
+    - [s''] is the result of annotating [s'] in [env];
+    - [otherwise'] is the result of annotating [otherwise] in [env];
+    - [catchers'] is the result of annotating [catchers] in [env];
+    - [new_s] is a try statement with statement [s''], catchers [catchers'] and block [otherwise'];
+    - [new_env] is [env].
 
     {4 Example}
 
   {3 TypingRule.SDebug}
 
     {4 Rule}
+    All of the following applies:
+    - [s] is a debug statement with expression [e];
+    - [_t_e, e'] is the result of annotating [e] in [env];
+    - [new_s] is a debug statement with expression [e'];
+    - [new_env] is [env].
```

Comment:
I do not think this should be in the spec. This is unspecified, and a very-asllib only syntax and code.

Probably once we fix the type-checking of the procedure `print`, all the `debug` statements could be converted in the ASL into `print` statements.